source/agora/script/Signature.d

```diff
@@ -142,39 +213,147 @@ unittest
         tx = the transaction to sign
         sig_hash = the `SigHash` to use
         input_idx = the associated input index we're signing for
+        output_idx = the associated output index we're signing for
 
     Returns:
         the challenge as a hash
 
 *******************************************************************************/
 
 public Hash getChallenge (in Transaction tx, in SigHash sig_hash = SigHash.All,
-    in ulong input_idx = 0) nothrow @safe
+    in ulong input_idx = 0, in ulong output_idx = 0) nothrow @safe
 {
+    import agora.script.Lock;
+    import std.exception : assumeWontThrow;
+
+    if (sig_hash != SigHash.All)
+    {
+        assert(input_idx < tx.inputs.length, "Input index is out of range");
+        if (sig_hash & SigHash.Single)
+            assert(output_idx < tx.outputs.length, "Output index is out of range");
+    }
+
+    Transaction dup = () @trusted { return *cast(Transaction*)&tx; }();
```

Comment:
We really need to get rid of this...

Reply:
Deep-dup thru ser/deser is not possible atm.
`TxBuilder` is signing invalid TXs while trying to calculate minimum fees and that triggers assertions in ctors.

https://github.com/bosagora/agora/blob/be48645a0c1e921bb562c5ed10873bf7ceccf3de/source/agora/utils/TxBuilder.d#L598